src/extensions/extensionsHelper.ts

```diff
@@ -503,52 +531,284 @@ export async function publishExtensionVersionFromLocalSource(args: {
+        { exit: 103 }
+      );
+    }
+  }
+
+  return { extensionSpec, notes };
+}
+
+/**
+ * Publishes an Extension version from a remote repo.
+ *
+ * @param publisherId the ID of the Publisher this Extension will be published under
+ * @param extensionId the ID of the Extension to be published
+ * @param repoUri the URI of the repo where this Extension's source exists
+ * @param sourceRef the commit hash, branch, or tag name in the repo to publish from
+ * @param extensionRoot the root directory that contains this Extension's source
+ * @param stage the release stage to publish
+ * @param nonInteractive whether to display prompts
+ * @param force whether to force confirmations
+ */
+export async function publishExtensionVersionFromRemoteRepo(args: {
+  publisherId: string;
+  extensionId: string;
+  repoUri: string;
+  sourceRef: string;
+  extensionRoot: string;
+  stage: ReleaseStage;
+  nonInteractive: boolean;
+  force: boolean;
+}): Promise<ExtensionVersion | undefined> {
+  const extensionRef = `${args.publisherId}/${args.extensionId}`;
+
+  // Check if Extension already exists.
+  let extension: Extension | undefined;
+  try {
+    extension = await getExtension(extensionRef);
+  } catch (err: any) {
+    // Silently fail and continue the publish flow if Extension not found.
+  }
+
+  // Prompt for repo URI and validate that it hasn't changed if previously set.
+  if (args.repoUri && !repoRegex.test(args.repoUri)) {
+    throw new FirebaseError("Repo URI follows this format: https://github.com/<user>/<repo>");
+  }
+  let repoUri = args.repoUri || extension?.repoUri;
+  if (!repoUri) {
+    repoUri = await promptForValidRepoURI();
+  }
+  if (extension?.repoUri) {
+    if (repoUri !== extension.repoUri) {
+      throw new FirebaseError(
+        `Repo URI '${clc.bold(args.repoUri)}' does not match repo URI '${clc.bold(
+          extension.repoUri!
+        )}' already associated with Extension ${clc.bold(
+          extensionRef
+        )}. Repo URI cannot be changed.`
+      );
+    } else {
+      logger.info(
+        `Extension ${clc.bold(extensionRef)} is published from ${clc.bold(extension?.repoUri)}.`
+      );
+    }
+  } else {
+    logger.info(
+      `\n${clc.red("Warning:")} You are about to associate repo URI ${clc.bold(
+        repoUri
+      )} with Extension ${clc.bold(
+        extensionRef
+      )}. This cannot be changed. All future verifiable versions must be published from this repo. ` +
+        `You can continue publishing unverifiable versions from local source.`
+    );
+    const confirmed = await confirm({
       nonInteractive: args.nonInteractive,
       force: args.force,
       default: false,
-    }))
-  ) {
-    return;
+    });
+    if (!confirmed) {
+      return;
+    }
+  }
+
+  // Prompt for Extension root and default to last root used.
+  let extensionRoot = args.extensionRoot;
+  let defaultRoot = "/";
+  if (!extensionRoot) {
```

Comment:
This logic should to be cleaned up - currently, it doesn't default to the previous versions root when noninteractive = true. I think it can be unnested a bit as well.

Reply:
Fixed the non-interactive case but can't see any way to unnest unless we do the HTTP call even if an argument has already been passed in to override it.